docs/naps/4-async-slicing.md

```diff
@@ -0,0 +1,636 @@
+For example, `Points.face_color` is indexed by the points that are visible in the current slice.
+
+#### Output state
+
+The output of slicing is typically layer-type specific, stored as state
+on the layer, and mostly consumed by the corresponding vispy layer.
+    
+- `_ImageBase`
+    - `_slice: ImageSlice`, contains a loader, and the sliced image and thumbnail
+        - much complexity encapsulated here and other related classes like `ImageSliceData`.
+- `Points`
+    - `__indices_view: Array[int, (M,)]`, indices of points (i.e. rows of `data`) that are in the current slice.
+        - many private properties derived from this (e.g. `_indices_view`, `_view_data`).
+    - `_view_size_scale: Union[float, Array[float, (M,)]]`, used with thick slices of points `_view_size` to make out of slice points appear smaller.
+- `Shapes`
+    - `_data_view: ShapeList`, stores all shapes' data.
+		- `_mesh: Mesh`, stores concatenated meshes of all shapes.
+			- `displayed_triangles: Array[int, (M, 3)]`, triangles to be drawn.
+			- `displayed_triangles_colors: Array[float, (M, 4)]`, per triangle color.
+- `Vectors`
+    - `_view_indices: Array[int, (-1)]`, indices of vectors that are in the current slice.
+        - lots of private properties derived from this (e.g. `_view_face_color`, `_view_faces`).
+
+The vispy layers also read other state from their corresponding layer.
+In particular they read `Layer._transforms` to produce a transform that can be used to properly
+represent the layer in the vispy scene coordinate system.
+
+## Detailed description
+
+The following diagram shows the new proposed approach to slicing layers asynchronously.
+
+![](https://raw.githubusercontent.com/andy-sweet/napari-diagrams/main/napari-slicing-async-calls.drawio.svg)
+
+As with the existing synchronous slicing design, the `Dims.events.current_step` event
+is the shared starting point. In the new approach, we pass `ViewerModel.layers` through
+to the newly defined `LayerSlicer`, which makes a slice request for each layer on the main thread.
+This request is processed asynchronously on a dedicated thread for slicing, while the main thread
+returns quickly to the Qt main event loop, allowing napari to keep responding to other updates
+and interactions. When all the layers have generated slice responses on the slicing thread,
+the `slice_ready` event is emitted. That triggers `QtViewer.on_slice_ready` to be executed on
+the main thread, so that the underlying `QWidgets` can be safely updated.
+
+The rest of this section defines some new types to encapsulate state that is critical
+to slicing and some new methods that redefine the core logic of slicing.
+
+### Request and response
+
+First, we introduce a request to encapsulate the required input to slicing.
+
+```python
+class LayerSliceRequest:
+    data: ArrayLike
+    world_to_data: Transform
+    point: Tuple[float, ...]
+    dims_displayed: Tuple[int, ...]
+    dims_not_displayed: Tuple[int, ...]
```

Comment:
I wonder if we should create a dedicated `Slice` object that includes all the includes all the encapsulated dims and camera info to determine what part of the data to show. `Slice` could then be an additional argument to a slicing function. Thoughts @jni - I know you don't want a proliferation of classes, but to me that seems like a helpful one

Reply:
Yeah, I like this idea and I think @nclack has a similar one: https://github.com/napari/napari/pull/4892#discussion_r935877117

Would be great to fold a lot of base layer's request state into a world bounding box.